XmlParser/src/ua/goit/alg/parser/StateMachine.java

```diff
@@ -0,0 +1,270 @@
+package ua.goit.alg.parser;
+
+public class StateMachine {
+  private static State state = State.INIT;
+  private XmlParserImpl xmlParser;
+  private TagCreator creator = new TagCreator();
+
+  public StateMachine(XmlParserImpl xmlParser) {
+    this.xmlParser = xmlParser;
+  }
+
+  public void next(char c) {
+    state = state.next(c, creator, xmlParser);
+  }
+
+  public enum State {
+    INIT {
+      @Override
+      public State next(char c, TagCreator creator, XmlParserImpl xmlParser) {
+        State result = INVALID_END;
+        if (c == '<') {
+          result = INIT;
+        }
+        if (c == '?') {
+          result = VALID;
+        }
+        System.out.println("INIT");
+        return result;
+      }
+    },
+    VALID {
+      @Override
+      public State next(char c, TagCreator creator, XmlParserImpl xmlParser) {
+        State result = INVALID_END;
+        if (c != '>') {
+          result = VALID;
+        } else {
+          result = START;
+        }
+
+        System.out.println("VALID " + c);
+        return result;
+      }
+    },
+    START {
+      @Override
+      public State next(char c, TagCreator creator, XmlParserImpl xmlParser) {
+        State result = INVALID_END;
+        if (c == '<') {
+          result = OPEN_TAG;
+        }
+
+        if (Character.isWhitespace(c)) {
+          result = END;
+        }
+
+        System.out.println("START " + c);
+        return result;
+      }
+    },
+    OPEN_TAG {
+      @Override
+      public State next(char c, TagCreator creator, XmlParserImpl xmlParser) {
+        State result = INVALID_END;
+        if (Character.isLetter(c)) {
+          creator.appendTagName(c);
+          result = TAG_NAME;
+        }
+
+        if (c == '/') {
+          result = CLOSE_TAG;
+        }
+
+        System.out.println("OPEN_TAG " + c);
+        return result;
+      }
+    },
+    TAG_NAME {
+      @Override
+      public State next(char c, TagCreator creator, XmlParserImpl xmlParser) {
+        State result = INVALID_END;
+        if (Character.isLetterOrDigit(c)) {
+          creator.appendTagName(c);
+          result = TAG_NAME;
```

Comment:
I'd suggest to return in ifs and sort them vice versa. it's easier to reason when amount of possible things which could happen decreases with each new line of code